rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -40,127 +102,227 @@ class SingleStateFeaturizer:
+
+        # FIXME: If this is re-instantiated for a policy that we want to fine-tune
+        # then the information to use a regex interpreter will be lost unless
+        # the  NLU Interpreter is still not present during re-training....
+        # Is this a problem?
+
+    def prepare(
+        self, domain: Domain, interpreter: Optional[NaturalLanguageInterpreter]
+    ) -> None:
+        """
+        TODO: all of this looks like it could be done in the init....
+        TODO/FIXME: the item2index_mappings are also needed during inference, which
+          is why it is very confusing this is computed in "prepare_for_training" only...
+        """
+        # TODO: the following this was if...: = True before, is this a problem?
+        self.use_regex_interpreter = isinstance(interpreter, RegexInterpreter)
+        self.item2index_mappings = {
+            key: item2index_mapping(items)
+            for key, items in [
+                (INTENT, domain.intents),
+                (ACTION_NAME, domain.action_names_or_texts),
+                (ENTITIES, domain.entity_states),
+                (SLOTS, domain.slot_states),
+                (ACTIVE_LOOP, domain.form_names),
+            ]
+        }
+        self.action_texts = domain.action_texts
+
+    def featurize_state(
+        self, state: State, interpreter: Optional[NaturalLanguageInterpreter]
+    ) -> Dict[Text, List["Features"]]:
+        """Encode the given state with the help of the given interpreter.
+
+        Args:
+            state: The state to encode
+            interpreter: The interpreter used to encode the state
 
         Returns:
-            Tag to index mapping.
+            A dictionary of state_type to list of features.
         """
-        if ENTITIES not in self._default_feature_states:
-            return []
+        # Note: This method is called during both prediction and training.
+        # `self._use_regex_interpreter == True` means that core was trained
+        # separately. Therefore we have to replace the passed interpreter
+        # (which is based on some trained nlu model)  with a default
+        # RegexInterpreter (i.e. same as during training) to make sure
+        # that prediction and train time features are the same
 
-        if bilou_tagging:
-            tag_id_index_mapping = {
-                f"{prefix}{tag}": idx_1 * len(BILOU_PREFIXES) + idx_2 + 1
-                for tag, idx_1 in self._default_feature_states[ENTITIES].items()
-                for idx_2, prefix in enumerate(BILOU_PREFIXES)
-            }
-        else:
-            tag_id_index_mapping = {
-                tag: idx + 1  # +1 to keep 0 for the NO_ENTITY_TAG
-                for tag, idx in self._default_feature_states[ENTITIES].items()
-            }
+        # FIXME/TODO:
+        # 1. Would be cool if users would be notified that they're not training
+        #    the policy based on what their NLU pipeline returns.... (cf.
+        #    featurize_via_interpreter)
```

Comment:
Agreed

Reply:
I was really confused whether the wrong interpreter might be used by accident. But I guess the reason there's no mismatch (in practice now) is in the .persist of the TrackerFeaturizer which will pickles this StateFeaturizer as a wholeand preserve the "use_regex..." flag.